prov/efa/src/efa_domain.c

```diff
@@ -230,19 +233,31 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 	}
 
 	efa_domain->mr_local = ofi_mr_local(info);
-	if (EFA_EP_TYPE_IS_DGRAM(info) && !efa_domain->mr_local) {
-		EFA_WARN(FI_LOG_EP_DATA, "dgram require FI_MR_LOCAL, but application does not support it\n");
+	if ((EFA_EP_TYPE_IS_DGRAM(info) || EFA_EP_TYPE_IS_DIRECT(info)) && !efa_domain->mr_local) {
+		EFA_WARN(FI_LOG_EP_DATA, "EFA direct and dgram require FI_MR_LOCAL, but application does not support it\n");
 		ret = -FI_ENODATA;
 		goto err_free;
 	}
 
-	err = efa_domain_init_device_and_pd(efa_domain, info->domain_attr->name, info->ep_attr->type);
+	err = efa_domain_init_device_and_pd(efa_domain, info->domain_attr->name, efa_ep_type);
 	if (err) {
 		ret = err;
 		goto err_free;
 	}
 
-	efa_domain->info = fi_dupinfo(EFA_EP_TYPE_IS_RDM(info) ? efa_domain->device->rdm_info : efa_domain->device->dgram_info);
+	switch (efa_ep_type) {
+		case EFA_EP_RDM:
+			efa_domain->info = fi_dupinfo(efa_domain->device->rdm_info);
```

Comment:
there shouldn't be a difference between `device->rdm_info` and `device->direct_info`. If we have anything today in `device->rdm_info` that doesn't reflect device capabilities, it is wrong and we need to fix.

Reply:
`fabric_attr->name` is different. `prov_info->mode` for efa-direct requires `FI_CONTEXT2`. `prov_info->tx_attr->msg_order` and `prov_info->rx_attr->msg_order` are different. HMEM capabilities are also different. `device->rdm_info` will support HMEM even without p2p.

We can do all of this in the user_info step but that's not how it is done now.

Reply:
Discussed offline. Refactoring to move logic to user_info will be done later.